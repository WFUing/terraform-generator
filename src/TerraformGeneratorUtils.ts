import replaceString from 'replace-string';
import { TerraformVersion, Block, Argument, Map } from '.';

export type ArgumentsToStringFn = (version: TerraformVersion, key: string, value: any) => string;

export default class TerraformGeneratorUtils {

  private static readonly escapeChars = [
    ['"', '&tfgquot;']
  ];

  static escape(str: string): string {
    this.escapeChars.forEach(char => str = replaceString(str, char[0], char[1]));
    return str;
  }

  static unescape(str: string): string {
    this.escapeChars.slice().reverse().forEach(char => str = replaceString(str, char[1], char[0]));
    return str;
  }

  static argumentsToString(version: TerraformVersion, args: Record<string, any>, customArgumentToString?: ArgumentsToStringFn): string {
    let str = '';
    for (const key in args) {
      if (customArgumentToString) {
        str += customArgumentToString(version, key, args[key]);
      } else {
        str += this.argumentToString(version, key, args[key]);
      }
    }
    return str;
  }

  static isObjectArgument(value: any): boolean {
    if (['string', 'number', 'boolean'].indexOf(typeof value) >= 0
      || value instanceof Block || value instanceof Argument || value instanceof Map) {
      return false;

    } else if (typeof value === 'object') {
      return true;

    } else {
      throw new Error(`Invalid value: ${value}`);
    }
  }

  static argumentToString(version: TerraformVersion, key: string, value: any): string {
    try {
      if (value == null) {
        return '';
      }

      if (key.indexOf(' ') > 0) {
        key = `"${key}"`;
      }

      let operator = ' = ';
      let isObjectArray = false;

      if (Array.isArray(value)) {
        if (value.length === 0 || this.isObjectArgument(value[0])) {
          value = value.filter(element => element != null);
          operator = ' ';
          isObjectArray = true;
        }
      } else {
        if (this.isObjectArgument(value)) {
          operator = ' ';
        }
      }

      if (isObjectArray) {
        let str = '';
        if (Array.isArray(value)) {
          value.forEach(element => {
            str += `${key}${operator}${this.argumentValueToString(version, element)}\n`;
          });
        }
        return str;

      } else {
        return `${key}${operator}${this.argumentValueToString(version, value)}\n`;
      }

    } catch (err) {
      throw new Error(`Invalid value: ${key} = ${value}`);
    }
  }

  static argumentValueToString(version: TerraformVersion, value: any): string {
    if (value == null) {
      return null;

    } else if (value instanceof Block) {
      return this.argumentValueToString(version, value.asArgument());

    } else if (value instanceof Argument) {
      if (version && version === '0.11' && !value.asIs) {
        return `"\${${value.toTerraform()}}"`;
      } else {
        return value.toTerraform();
      }

    } else if (value instanceof Map) {
      return this.argumentValueToString(version, value.arguments);

    } else if (['string', 'number', 'boolean'].indexOf(typeof value) >= 0) {
      return JSON.stringify(value);

    } else if (typeof value === 'object') {
      if (Array.isArray(value)) {
        let str = '[\n';
        value.forEach((element, i) => {
          str += `${this.argumentValueToString(version, element)}${i < value.length - 1 ? ',' : ''}\n`;
        });
        str += ']';
        return str;

      } else {
        let str = '{\n';
        for (const key in value) {
          str += this.argumentToString(version, key, value[key]);
        }
        str += '}';
        return str;
      }

    } else {
      throw new Error(`Invalid value: ${value}`);
    }
  }

}
