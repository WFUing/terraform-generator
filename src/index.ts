export { BlockArgs } from './utils';

export { TerraformGenerator, WriteOptions } from './TerraformGenerator';

export {
  Block,
  Backend,
  Comment,
  Data,
  Import, ImportArgs,
  Locals,
  Module, ModuleArgs,
  Output, OutputArgs,
  Provider,
  Provisioner, ProvisionerArgs, ProvisionerType,
  Resource, ResourceToDataOptions,
  Variable, VariableArgs
} from './blocks';

export {
  Argument, arg,
  Attribute, attr,
  Function, fn,
  Heredoc, heredoc
} from './arguments';

export {
  List, list,
  Map, map
} from './types';
