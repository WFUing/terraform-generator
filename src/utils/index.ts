export * from './types';
export * from './Util';
