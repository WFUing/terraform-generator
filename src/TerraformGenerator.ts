import shell from 'shelljs';
import child_process from 'child_process';
import fs from 'fs';
import path from 'path';
import { Block, Resource, Data, Module, Output, Provider, Variable, Backend, ResourceToDataOptions } from '.';
import TerraformGeneratorUtils from './TerraformGeneratorUtils';

export type TerraformVersion = '0.11' | '0.12';

/**
 * @param version Terraform version
 */
export interface TerraformGeneratorOptions {
  version: TerraformVersion;
}

/**
 * @param dir directoty, default = .
 * @param filename Terraform filename, must ends with .tf, default = terraform.tf
 * @param format use 'terraform fmt' to format the configuration, Terraform must be installed, default = false
 */
export interface WriteOptions {
  dir?: string;
  filename?: string;
  format?: boolean;
}

export default class TerraformGenerator {

  private readonly options: TerraformGeneratorOptions;
  private readonly arguments: Record<string, any>;
  private readonly blocks: Block[] = [];

  /**
   * Construct Terraform generator.
   * Refer to Terraform documentation on what can be put as arguments.
   * 
   * @param options options
   * @param args arguments
   */
  constructor(options: TerraformGeneratorOptions, args?: Record<string, any>) {
    this.options = options;
    this.arguments = args;
  }

  /**
   * Generate Terraform configuration as string.
   */
  generate(): string {
    let str = '';

    if (this.arguments || this.blocks.filter(block => block instanceof Backend).length > 0) {
      str += 'terraform {\n';
      str += TerraformGeneratorUtils.argumentsToString(this.options.version, this.arguments);
      this.blocks.forEach(block => {
        if (block instanceof Backend) {
          str += block.toTerraform(this.options.version);
        }
      });
      str += '}\n\n';
    }

    this.blocks.forEach(block => {
      if (!(block instanceof Backend)) {
        str += block.toTerraform(this.options.version);
      }
    });

    return str;
  }

  /**
   * Write Terraform configuration to a file.
   * 
   * @param options options
   */
  write(options?: WriteOptions): void {
    if (!options) {
      options = {};
    }
    if (!options.dir) {
      options.dir = '.';
    }
    if (!options.filename) {
      options.filename = 'terraform.tf';
    }
    if (!options.filename.endsWith('.tf')) {
      options.filename += '.tf';
    }

    shell.mkdir('-p', options.dir);
    fs.writeFileSync(path.join(options.dir, options.filename), this.generate());

    if (options.format) {
      child_process.execSync('terraform fmt', { cwd: options.dir });
    }
  }

  /**
   * Add blocks into Terraform.
   * 
   * @param blocks blocks
   */
  addBlocks(...blocks: Block[]): TerraformGenerator {
    blocks.forEach(block => this.blocks.push(block));
    return this;
  }

  /**
   * Add provider into Terraform.
   * Refer to Terraform documentation on what can be put as type & arguments.
   * 
   * @param type type
   * @param args arguments
   */
  provider(type: string, args?: Record<string, any>): Provider {
    const block = new Provider(type, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add resource into Terraform.
   * Refer to Terraform documentation on what can be put as type & arguments.
   * 
   * @param type type
   * @param name name
   * @param args arguments
   */
  resource(type: string, name: string, args?: Record<string, any>): Resource {
    const block = new Resource(type, name, args);
    this.addBlocks(block);
    return block;
  }
  /**
   * Convert resource into data source and add it into Terraform.
   * 
   * @param resource resource
   * @param options options
   * @param argNames names of resource arguments to be converted into data source arguments;
   * use array for name mapping, position 0 = original resource's argument name, position 1 = mapped data source's argument name
   * @param args extra arguments
   */
  dataFromResource(resource: Resource, options: ResourceToDataOptions, argNames: (string | [string, string])[], args?: Record<string, any>): Data {
    const block = resource.toData(options, argNames, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add data source into Terraform.
   * Refer to Terraform documentation on what can be put as type & arguments.
   * 
   * @param type type
   * @param name name
   * @param args arguments
   */
  data(type: string, name: string, args?: Record<string, any>): Data {
    const block = new Data(type, name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add module into Terraform.
   * Refer to Terraform documentation on what can be put as arguments.
   * 
   * @param name name
   * @param args arguments
   */
  module(name: string, args?: Record<string, any>): Module {
    const block = new Module(name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add output into Terraform.
   * Refer to Terraform documentation on what can be put as arguments.
   * 
   * @param name name
   * @param args arguments
   */
  output(name: string, args?: Record<string, any>): Output {
    const block = new Output(name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add provider into Terraform.
   * Refer to Terraform documentation on what can be put as arguments.
   * 
   * @param name name
   * @param args arguments
   */
  variable(name: string, args?: Record<string, any>): Variable {
    const block = new Variable(name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add backend into Terraform.
   * Refer to Terraform documentation on what can be put as type & arguments.
   * 
   * @param type type
   * @param args arguments
   */
  backend(type: string, args?: Record<string, any>): Backend {
    const block = new Backend(type, args);
    this.addBlocks(block);
    return block;
  }

}
