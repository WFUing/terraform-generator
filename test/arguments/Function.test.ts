import TerraformGeneratorUtils from '../../src/TerraformGeneratorUtils';
import { Function, fn } from '../../src';
import { attr } from '..';

test('Function invalid args', () => {
  expect(() => new Function(null)).toThrow();
  expect(() => new Function('fn', null)).toThrow();
});

test('Function', () => {
  expect(new Function('fn').toString()).toBe(TerraformGeneratorUtils.escape('fn()'));
  expect(new Function('fn', 'x', 'y', 'z').toString()).toBe(TerraformGeneratorUtils.escape('fn("x", "y", "z")'));
  expect(new Function('fn', 1, 2, 3).toString()).toBe(TerraformGeneratorUtils.escape('fn(1, 2, 3)'));
  expect(new Function('fn', true, false, true).toString()).toBe(TerraformGeneratorUtils.escape('fn(true, false, true)'));
  expect(new Function('fn', 'x', 2, true).toString()).toBe(TerraformGeneratorUtils.escape('fn("x", 2, true)'));
  expect(new Function('fn', 'x', attr).toString()).toBe(TerraformGeneratorUtils.escape('fn("x", type.name.attr)'));
  expect(new Function('fn', 'x', { a: 1, b: '2' }).toString()).toBe(TerraformGeneratorUtils.escape('fn("x", {\na = 1\nb = "2"\n})'));
});

test('fn', () => {
  expect(fn('fn').toString()).toBe(TerraformGeneratorUtils.escape('fn()'));
  expect(fn('fn', 'x', 'y', 'z').toString()).toBe(TerraformGeneratorUtils.escape('fn("x", "y", "z")'));
  expect(fn('fn', 1, 2, 3).toString()).toBe(TerraformGeneratorUtils.escape('fn(1, 2, 3)'));
  expect(fn('fn', true, false, true).toString()).toBe(TerraformGeneratorUtils.escape('fn(true, false, true)'));
  expect(fn('fn', 'x', 2, true).toString()).toBe(TerraformGeneratorUtils.escape('fn("x", 2, true)'));
  expect(fn('fn', 'x', attr).toString()).toBe(TerraformGeneratorUtils.escape('fn("x", type.name.attr)'));
  expect(fn('fn', 'x', { a: 1, b: '2' }).toString()).toBe(TerraformGeneratorUtils.escape('fn("x", {\na = 1\nb = "2"\n})'));
});
