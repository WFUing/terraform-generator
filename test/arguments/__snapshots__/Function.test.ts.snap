// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Function 1`] = `"fn()"`;

exports[`Function 2`] = `"fn(&tfgquot;x&tfgquot;, &tfgquot;y&tfgquot;, &tfgquot;z&tfgquot;)"`;

exports[`Function 3`] = `"fn(1, 2, 3)"`;

exports[`Function 4`] = `"fn(true, false, true)"`;

exports[`Function 5`] = `"fn(&tfgquot;x&tfgquot;, 2, true)"`;

exports[`Function 6`] = `"fn(&tfgquot;x&tfgquot;, type.name.attr)"`;

exports[`Function 7`] = `
"fn(&tfgquot;x&tfgquot;, {
a = 1
b = &tfgquot;2&tfgquot;
})"
`;

exports[`fn 1`] = `"fn()"`;

exports[`fn 2`] = `"fn(&tfgquot;x&tfgquot;, &tfgquot;y&tfgquot;, &tfgquot;z&tfgquot;)"`;

exports[`fn 3`] = `"fn(1, 2, 3)"`;

exports[`fn 4`] = `"fn(true, false, true)"`;

exports[`fn 5`] = `"fn(&tfgquot;x&tfgquot;, 2, true)"`;

exports[`fn 6`] = `"fn(&tfgquot;x&tfgquot;, type.name.attr)"`;

exports[`fn 7`] = `
"fn(&tfgquot;x&tfgquot;, {
a = 1
b = &tfgquot;2&tfgquot;
})"
`;
